File: scala/scalasearch/src/main/scala/scalasearch/SearchResult.scala
package scalasearch

import scalafind.Color.Color
import scalafind.{FileResult, FileResultFormatter, SortBy}

import scala.util.matching.Regex

case class SearchResult(searchPattern: Regex, file: Option[FileResult],
                        lineNum: Int, matchStartIndex: Int,
                        matchEndIndex: Int, line: Option[String],
                        linesBefore: Seq[String], linesAfter: Seq[String]) {

  def this(searchPattern: Regex, file: Option[FileResult], lineNum: Int,
           matchStartIndex: Int, matchEndIndex: Int, line: Option[String]) = {
    this(searchPattern, file, lineNum, matchStartIndex, matchEndIndex, line,
      Seq.empty[String], Seq.empty[String])
  }
  
  def compareBySearchFields(other: SearchResult): Int = {
    if (this.lineNum == other.lineNum) {
      if (this.matchStartIndex == other.matchStartIndex) {
        if (this.matchEndIndex < other.matchEndIndex) -1 else 1
      } else {
        if (this.matchStartIndex < other.matchStartIndex) -1 else 1
      }
    } else {
      if (this.lineNum < other.lineNum) -1 else 1
    }
  }

  def compareByPath(other: SearchResult, sortCaseInsensitive: Boolean): Int = {
    val fileCmp: Int = (this.file, other.file) match {
      case (Some(thisFile), Some(otherFile)) =>
        thisFile.compareByPath(otherFile, sortCaseInsensitive)
      case (_, _) => 0
    }
    if (fileCmp == 0) {
      compareBySearchFields(other)
    } else {
      fileCmp
    }
  }

  def beforeByPath(other: SearchResult, sortCaseInsensitive: Boolean): Boolean = {
    val cmp = compareByPath(other, sortCaseInsensitive)
    cmp < 0
  }

  def compareByName(other: SearchResult, sortCaseInsensitive: Boolean): Int = {
    val fileCmp: Int = (this.file, other.file) match {
      case (Some(thisFile), Some(otherFile)) =>
        thisFile.compareByName(otherFile, sortCaseInsensitive)
      case (_, _) => 0
    }
    if (fileCmp == 0) {
      compareBySearchFields(other)
    } else {
      fileCmp
    }
  }

  def beforeByName(other: SearchResult, sortCaseInsensitive: Boolean): Boolean = {
    val cmp = compareByName(other, sortCaseInsensitive)
    cmp < 0
  }

  def compareBySize(other: SearchResult, sortCaseInsensitive: Boolean): Int = {
    val fileCmp: Int = (this.file, other.file) match {
      case (Some(thisFile), Some(otherFile)) =>
        thisFile.compareBySize(otherFile, sortCaseInsensitive)
      case (_, _) => 0
    }
    if (fileCmp == 0) {
      compareBySearchFields(other)
    } else {
      fileCmp
    }
  }

  def beforeBySize(other: SearchResult, sortCaseInsensitive: Boolean): Boolean = {
    val cmp = compareBySize(other, sortCaseInsensitive)
    cmp < 0
  }

  def compareByType(other: SearchResult, sortCaseInsensitive: Boolean): Int = {
    val fileCmp: Int = (this.file, other.file) match {
      case (Some(thisFile), Some(otherFile)) =>
        thisFile.compareByType(otherFile, sortCaseInsensitive)
      case (_, _) => 0
    }
    if (fileCmp == 0) {
      compareBySearchFields(other)
    } else {
      fileCmp
    }
  }

  def beforeByType(other: SearchResult, sortCaseInsensitive: Boolean): Boolean = {
    val cmp = compareByType(other, sortCaseInsensitive)
    cmp < 0
  }

  def compareByLastMod(other: SearchResult, sortCaseInsensitive: Boolean): Int = {
    val fileCmp: Int = (this.file, other.file) match {
      case (Some(thisFile), Some(otherFile)) =>
        thisFile.compareByLastMod(otherFile, sortCaseInsensitive)
      case (_, _) => 0
    }
    if (fileCmp == 0) {
      compareBySearchFields(other)
    } else {
      fileCmp
    }
  }

  def beforeByLastMod(other: SearchResult, sortCaseInsensitive: Boolean): Boolean = {
    val cmp = compareByLastMod(other, sortCaseInsensitive)
    cmp < 0
  }
}

class SearchResultFormatter(val settings: SearchSettings) {
  val sepLen = 80
  var fileResultFormatter = new FileResultFormatter(settings.getFindSettings)

  private def formatLineWithColor(line: String): String = {
    var formattedLine = line
    settings.searchPatterns.flatMap(p => p.findFirstMatchIn(formattedLine)).take(1).foreach { m =>
      formattedLine = colorize(formattedLine, m.start, m.end, settings.lineColor)
    }
    formattedLine
  }

  val formatLine: String => String =
    if (settings.colorize) {
      formatLineWithColor
    } else {
      (line: String) => line
    }

  private def formatMatchWithColor(m: String): String = {
    colorize(m, 0, m.length, settings.lineColor)
  }

  val formatMatch: String => String =
    if (settings.colorize) {
      formatMatchWithColor
    } else {
      (m: String) => m
    }

  def format(result: SearchResult): String = {
    if (result.linesBefore.nonEmpty || result.linesAfter.nonEmpty) {
      multiLineFormat(result)
    } else {
      singleLineFormat(result)
    }
  }

  private def colorize(s: String, matchStartIndex: Int, matchEndIndex: Int, color: Color): String = {
    fileResultFormatter.colorize(s, matchStartIndex, matchEndIndex, color)
  }

  private def formatResultMatch(result: SearchResult): String = {
    result.line match {
      case Some(line) =>
        var matchStartIndex = result.matchStartIndex - 1
        var matchEndIndex = result.matchEndIndex - 1
        val matchLength = matchEndIndex - matchStartIndex

        var prefix = ""
        var suffix = ""
        var colorStartIndex = 0
        var colorEndIndex = matchLength

        if (matchLength > settings.maxLineLength) {
          prefix = if (matchStartIndex > 2) "..." else ""
          suffix = "..."
          colorStartIndex = prefix.length
          colorEndIndex = settings.maxLineLength - 3
          matchEndIndex = matchStartIndex + colorEndIndex
          matchStartIndex = matchStartIndex + colorStartIndex
        }

        val matchString = prefix + line.substring(matchStartIndex, matchEndIndex) + suffix

        if (settings.colorize) {
          colorize(matchString, colorStartIndex, colorEndIndex, settings.lineColor)
        } else {
          matchString
        }
      case None => ""
    }
  }

  private def leadingWhitespaceCount(s: String): Int = {
    var idx = 0
    while (idx < s.length && Character.isWhitespace(s.charAt(idx))) {
      idx += 1
    }
    idx
  }

  private def trailingWhitespaceCount(s: String): Int = {
    var idx = s.length - 1
    while (idx > 0 && Character.isWhitespace(s.charAt(idx))) {
      idx -= 1
    }
    s.length - 1 - idx
  }

  private def getLineIndices(lineStartIdx: Int, lineEndIdx: Int, matchStartIdx: Int, matchEndIdx: Int, maxIdx: Int, maxLength: Int): (Int, Int, Int, Int) = {
    if (maxLength == 0 || lineEndIdx - lineStartIdx == 0) {
      (0, 0, 0, 0)
    } else if (lineEndIdx - lineStartIdx < maxLength) {
      val (lsi, msi, mei) =
        if (lineStartIdx > 0) {
          (lineStartIdx - 1, matchStartIdx + 1, matchEndIdx + 1)
        } else {
          (lineStartIdx, matchStartIdx, matchEndIdx)
        }
      val lei = if (lineEndIdx - lsi < maxLength && lineEndIdx < maxIdx) lineEndIdx + 1 else lineEndIdx
      getLineIndices(lsi, lei, msi, mei, maxIdx, maxLength)
    } else {
      (lineStartIdx, lineEndIdx, matchStartIdx, matchEndIdx)
    }
  }

  private def formatResultLineWithMatch(result: SearchResult): String = {
    if (result.line.isEmpty || result.line.get.trim.isEmpty || settings.maxLineLength == 0) {
      ""
    } else {
      val line = result.line.get
      val matchLength = result.matchEndIndex - result.matchStartIndex
      val maxLineLength =
        if (settings.maxLineLength < 0) {
          result.line.get.length + 1
        } else {
          settings.maxLineLength
        }

      val lineStartIdx = leadingWhitespaceCount(line)
      val lineEndIdx = line.length - 1 - trailingWhitespaceCount(line)
      val trimmedLength = lineEndIdx - lineStartIdx

      val matchStartIdx = result.matchStartIndex - 1
      val matchEndIdx = matchStartIdx + matchLength

      val (lsi, lei, msi, mei) =
        if (maxLineLength > 0 && trimmedLength > maxLineLength) {
          getLineIndices(matchStartIdx, matchEndIdx, 0, matchLength, trimmedLength, maxLineLength)
        } else {
          (lineStartIdx, lineEndIdx+1, matchStartIdx, matchEndIdx)
        }

      val (prefix, lsi2) =
        if (lsi > 2) {
          ("...", lsi + 3)
        } else {
          ("", lsi)
        }
      val (suffix, lei2) =
        if (lei < (trimmedLength - 3)) {
          ("...", lei - 3)
        } else {
          ("", lei)
        }

      val formatted = prefix + line.substring(lsi2, lei2) + suffix

      if (settings.colorize) {
        colorize(formatted, msi, mei, settings.lineColor)
      } else {
        formatted
      }
    }
  }

  private def formatResultLine(result: SearchResult): String = {
    if (result.line.isEmpty || result.line.get.trim.isEmpty || settings.maxLineLength == 0) {
      ""
    } else if (settings.maxLineLength > 0 && result.matchEndIndex - result.matchStartIndex > settings.maxLineLength) {
      formatResultMatch(result)
    } else {
      formatResultLineWithMatch(result)
    }
  }

  def singleLineFormat(result: SearchResult): String = {
    val filepath = if (result.file.isDefined) fileResultFormatter.formatFileResult(result.file.get) else "<text>"
    val matchString =
      if (result.lineNum > 0) {
        ": %d: [%d:%d]: %s".format(result.lineNum, result.matchStartIndex, result.matchEndIndex,
          formatResultLine(result))
      } else {
        " matches at [%d:%d]".format(result.matchStartIndex, result.matchEndIndex)
      }
    filepath + matchString
  }

  def lineNumPadding(result: SearchResult): Int = {
    val maxLineNum = result.lineNum + result.linesAfter.length
    "%d".format(maxLineNum).length
  }

  val newLineChars = Set('\n', '\r')

  def trimNewLines(s:String): String = {
    if (s == "") {
      ""
    } else if (newLineChars.contains(s.last)) {
      trimNewLines(s.init)
    } else {
      s
    }
  }

  def multiLineFormat(result: SearchResult): String = {
    result.line match {
      case Some(resultLine) =>
        val sb = new StringBuilder
        val filepath = if (result.file.isDefined) fileResultFormatter.formatFileResult(result.file.get) else "<text>"
        sb.append("%s\n%s: %d: [%d:%d]\n%s\n".format("=" * sepLen, filepath,
          result.lineNum, result.matchStartIndex, result.matchEndIndex, "-" * sepLen))
        val lineFormat = " %1$" + lineNumPadding(result) + "d | %2$s\n"
        var currentLineNum = result.lineNum
        if (result.linesBefore.nonEmpty) {
          currentLineNum -= result.linesBefore.length
          result.linesBefore.foreach { lineBefore =>
            sb.append(" " + lineFormat.format(currentLineNum, trimNewLines(lineBefore)))
            currentLineNum += 1
          }
        }
        var line = trimNewLines(resultLine)
        if (settings.colorize) {
          line = colorize(line, result.matchStartIndex - 1, result.matchEndIndex - 1, settings.lineColor)
        }
        sb.append(">" + lineFormat.format(result.lineNum, line))
        if (result.linesAfter.nonEmpty) {
          currentLineNum += 1
          result.linesAfter.foreach { lineAfter =>
            sb.append(" " + lineFormat.format(currentLineNum, trimNewLines(lineAfter)))
            currentLineNum += 1
          }
        }
        sb.toString()

      case None => ""
    }
  }
}

class SearchResultSorter(val settings: SearchSettings) {
  private def getSearchResultComparator: (SearchResult, SearchResult) => Boolean = {
    if (settings.sortDescending) {
      settings.sortBy match {
        case SortBy.FileName => (sr1: SearchResult, sr2: SearchResult) => sr2.beforeByName(sr1, settings.sortCaseInsensitive)
        case SortBy.FileSize => (sr1: SearchResult, sr2: SearchResult) => sr2.beforeBySize(sr1, settings.sortCaseInsensitive)
        case SortBy.FileType => (sr1: SearchResult, sr2: SearchResult) => sr2.beforeByType(sr1, settings.sortCaseInsensitive)
        case SortBy.LastMod => (sr1: SearchResult, sr2: SearchResult) => sr2.beforeByLastMod(sr1, settings.sortCaseInsensitive)
        case _ => (sr1: SearchResult, sr2: SearchResult) => sr2.beforeByPath(sr1, settings.sortCaseInsensitive)
      }
    } else {
      settings.sortBy match {
        case SortBy.FileName => (sr1: SearchResult, sr2: SearchResult) => sr1.beforeByName(sr2, settings.sortCaseInsensitive)
        case SortBy.FileSize => (sr1: SearchResult, sr2: SearchResult) => sr1.beforeBySize(sr2, settings.sortCaseInsensitive)
        case SortBy.FileType => (sr1: SearchResult, sr2: SearchResult) => sr1.beforeByType(sr2, settings.sortCaseInsensitive)
        case SortBy.LastMod => (sr1: SearchResult, sr2: SearchResult) => sr1.beforeByLastMod(sr2, settings.sortCaseInsensitive)
        case _ => (sr1: SearchResult, sr2: SearchResult) => sr1.beforeByPath(sr2, settings.sortCaseInsensitive)
      }
    }
  }

  def sort(searchResults: Seq[SearchResult]): Seq[SearchResult] = {
    val searchResultComparator = getSearchResultComparator
    searchResults.sortWith(searchResultComparator)
  }
}

File: elixir/exsearch/lib/searchresult.ex
defmodule ExSearch.SearchResult do
  @moduledoc """
  Documentation for `ExFind.FileResult`.
  """

  defstruct pattern: "", file: nil, line_num: 0, line: "", lines_before: [], lines_after: [], match_start_index: 0, match_end_index: 0

  def new(pattern, file, line_num, line, lines_before, lines_after, match_start_index, match_end_index) do
    __struct__(
      [
        pattern: pattern,
        file: file,
        line_num: line_num,
        line: line,
        lines_before: lines_before,
        lines_after: lines_after,
        match_start_index: match_start_index,
        match_end_index: match_end_index
      ]
    )
  end

  def new(args), do: __struct__(args)
end

defmodule ExSearch.SearchResultFormatter do
  @moduledoc """
  Documentation for `ExSearch.SearchResultFormatter`.
  """

  alias ExFind.FileResultFormatter

  defstruct [:settings, :file_formatter]

  def new(settings) do
    __struct__([
      settings: settings,
      file_formatter: FileResultFormatter.new(settings)
    ])
  end

  def format_result_match(formatter, search_result) do
    match_start_idx = search_result.match_start_index - 1
    match_end_idx = search_result.match_end_index - 1
    match_length = match_end_idx - match_start_idx

    {prefix, suffix, color_start_idx, color_end_idx, match_start_idx, match_end_idx} =
      if match_length > formatter.settings.max_line_length do
        prefix = if match_start_idx > 2, do: "...", else: ""
        suffix = "..."
        color_start_idx = String.length(prefix)
        color_end_idx = formatter.settings.max_line_length - 3
        match_end_idx = match_start_idx + color_end_idx
        match_start_idx = match_start_idx + color_start_idx
        {prefix, suffix, color_start_idx, color_end_idx, match_start_idx, match_end_idx}
      else
        {"", "", 0, match_end_idx, match_start_idx, match_end_idx}
      end

    match_string = prefix <> String.slice(search_result.line, match_start_idx, match_end_idx - match_start_idx) <> suffix
    if formatter.settings.colorize do
      FileResultFormatter.colorize(match_string, color_start_idx, color_end_idx, formatter.settings.line_color)
    else
      match_string
    end
  end

  def get_indices_for_max_line_length(line_start_idx, line_end_idx, match_start_idx, match_end_idx, line_length, max_line_length) do
    if (line_end_idx - line_start_idx) < max_line_length do
      {lsi, msi, mei} =
        if line_start_idx > 0 do
          {line_start_idx - 1, match_start_idx + 1, match_end_idx + 1}
        else
          {line_start_idx, match_start_idx, match_end_idx}
        end
      lei =
        if (line_end_idx - lsi) < max_line_length && line_end_idx < line_length do
          line_end_idx + 1
        else
          line_end_idx
        end
      get_indices_for_max_line_length(lsi, lei, msi, mei, line_length, max_line_length)
    else
      {line_start_idx, line_end_idx, match_start_idx, match_end_idx}
    end
  end

  def format_result_line(formatter, search_result) do
    if (search_result.line == "" || formatter.settings.max_line_length == 0) do
      ""
    else
      if formatter.settings.max_line_length > 0 && (search_result.match_end_index - search_result.match_start_index) > formatter.settings.max_line_length do
        format_result_match(formatter, search_result)
      else
        line_len = String.length(search_result.line)
        line_start_idx = line_len - String.length(String.trim_leading(search_result.line))
        line_end_idx = line_len - 1 - (line_len - String.length(String.trim_trailing(search_result.line)))
        match_length = search_result.match_end_index - search_result.match_start_index
        match_start_idx = search_result.match_start_index - 1 - line_start_idx
        match_end_idx = match_start_idx + match_length
        trimmed_length = line_end_idx - line_start_idx
        {line_start_idx, line_end_idx, match_start_idx, match_end_idx} =
          if formatter.settings.max_line_length > 0 && trimmed_length > formatter.settings.max_line_length do
            get_indices_for_max_line_length(search_result.match_start_index - 1, search_result.match_end_index - 1, 0, match_length, trimmed_length, formatter.settings.max_line_length)
          else
            {line_start_idx, line_end_idx + 1, match_start_idx, match_end_idx}
          end
        {prefix, line_start_idx} = if line_start_idx > 2, do: {"...", line_start_idx + 3}, else: {"", line_start_idx}
        {suffix, line_end_idx} = if line_end_idx < (trimmed_length - 2), do: {"...", line_end_idx - 3}, else: {"", line_end_idx}
        formatted = prefix <> String.slice(search_result.line, line_start_idx, line_end_idx - line_start_idx) <> suffix
        if formatter.settings.colorize do
          FileResultFormatter.colorize(formatted, match_start_idx, match_end_idx, formatter.settings.line_color)
        else
          formatted
        end
      end
    end
  end

  def format_lines_with_nums(lines, start_line_num, max_line_num) do
    line_num_padding = String.length("#{max_line_num}")
    lines_with_line_nums = Enum.with_index(lines, start_line_num)
    Enum.map(lines_with_line_nums, fn {line, line_num} ->
      line_num_str = String.pad_leading("#{line_num}", line_num_padding + 2, " ")
      "#{line_num_str} | #{line}"
    end)
  end

  def multi_line_format(formatter, search_result) do
    max_line_num = search_result.line_num + length(search_result.lines_after)
    case StringIO.open("", [], fn out ->
      match_index_range = "[#{search_result.match_start_index}:#{search_result.match_end_index}]"
      file_path = FileResultFormatter.format_file_result(formatter.file_formatter, search_result.file)
      file_path_line = "#{file_path}: #{search_result.line_num}: #{match_index_range}"
      IO.write(out, String.duplicate("=", 80) <> "\n")
      IO.write(out, "#{file_path_line}\n")
      IO.write(out, String.duplicate("-", 80) <> "\n")
      if search_result.lines_before != [] do
        format_lines_with_nums(search_result.lines_before, search_result.line_num - length(search_result.lines_before), max_line_num)
        |> Enum.each(fn line -> IO.write(out, "#{line}\n") end)
      end
      matching_line =
        if formatter.settings.colorize do
          FileResultFormatter.colorize(search_result.line, search_result.match_start_idx - 1, search_result.match_end_idx - 1, formatter.settings.line_color)
        else
          search_result.line
        end
      IO.write(out, "> #{search_result.line_num} | #{matching_line}\n")
      if search_result.lines_after != [] do
        format_lines_with_nums(search_result.lines_after, search_result.line_num + 1, max_line_num)
        |> Enum.each(fn line -> IO.write(out, "#{line}\n") end)
      end
      StringIO.contents(out)
    end) do
      {:ok, {"", multi_line_result}} -> multi_line_result
      _ -> ""
    end
  end

  def single_line_format(formatter, search_result) do
    match_range = "[#{search_result.match_start_index}:#{search_result.match_end_index}]"
    line_result = case {search_result.line_num, search_result.line} do
      {0, ""} -> " matches at #{match_range}"
      _ ->
        line = format_result_line(formatter, search_result)
        ": #{search_result.line_num}: #{match_range}: #{line}"
    end
    FileResultFormatter.format_file_result(formatter.file_formatter, search_result.file) <> line_result
  end

  def format(formatter, search_result) do
    if formatter.settings.lines_before > 0 or formatter.settings.lines_after > 0 do
      multi_line_format(formatter, search_result)
    else
      single_line_format(formatter, search_result)
    end
  end
end

defmodule ExSearch.SearchResultSorter do
  @moduledoc """
  Documentation for `ExSearch.SearchResultSorter`.
  """

  def get_file_path_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {String.downcase(r.file.path), String.downcase(r.file.name), r.line_num, r.match_start_index, r.match_end_index} end
    else
      fn r -> {r.file.path, r.file.name, r.line_num, r.match_start_index, r.match_end_index} end
    end
  end

  def get_file_name_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {String.downcase(r.file.name), String.downcase(r.file.path), r.line_num, r.match_start_index, r.match_end_index} end
    else
      fn r -> {r.file.name, r.file.path, r.line_num, r.match_start_index, r.match_end_index} end
    end
  end

  def get_file_size_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {r.file.file_size, String.downcase(r.file.path), String.downcase(r.file.name), r.line_num, r.match_start_index, r.match_end_index} end
    else
      fn r -> {r.file.file_size, r.file.path, r.file.name, r.line_num, r.match_start_index, r.match_end_index} end
    end
  end

  def get_file_type_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {r.file.file_type, String.downcase(r.file.path), String.downcase(r.file.name), r.line_num, r.match_start_index, r.match_end_index} end
    else
      fn r -> {r.file.file_type, r.file.path, r.file.name, r.line_num, r.match_start_index, r.match_end_index} end
    end
  end

  def get_last_mod_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {r.file.last_mod, String.downcase(r.file.path), String.downcase(r.file.name), r.line_num, r.match_start_index, r.match_end_index} end
    else
      fn r -> {r.file.last_mod, r.file.path, r.file.name, r.line_num, r.match_start_index, r.match_end_index} end
    end
  end

  def get_search_result_mapper(settings) do
    case settings.sort_by do
      :file_name -> get_file_name_mapper(settings)
      :file_size -> get_file_size_mapper(settings)
      :file_type -> get_file_type_mapper(settings)
      :last_mod  -> get_last_mod_mapper(settings)
      _          -> get_file_path_mapper(settings)
    end
  end

  def sort(settings, results) do
    search_result_mapper = get_search_result_mapper(settings)
    direction = if settings.sort_descending, do: :desc, else: :asc
    Enum.sort_by(results, search_result_mapper, direction)
  end
end
